### PR TITLE
[ISSUE-1359] reconnection handling for openpaas contact consumer

### PR DESCRIPTION
**Pull request by florentos17**
tackling [issue 1359](https://github.com/linagora/tmail-backend/issues/1359)

### CODE REVIEW

File: tmail-backend/tmail-third-party/openpaas/src/main/java/com/linagora/tmail/OpenPaasModule.java [hunk 1]
```diff
@@ -97,4 +98,9 @@ public ReactorRabbitMQChannelPool provideReactorRabbitMQChannelPool(
         channelPool.start();
         return channelPool;
     }
+
+    @ProvidesIntoSet
+    SimpleConnectionPool.ReconnectionHandler provideReconnectionHandler(@Named(OPENPAAS_INJECTION_KEY) OpenPaasContactsConsumer contactsConsumer) {
+        return new OpenPaasContactsConsumerReconnectionHandler(contactsConsumer);
+    }
```
Comment (florentos17) [inline tmail-backend/tmail-third-party/openpaas/src/main/java/com/linagora/tmail/OpenPaasModule.java:105]: not so sure about this, i tried to do the same as what exists.

should i add some kind of
```
@Override
    protected void configure() {
        Multibinder<SimpleConnectionPool.ReconnectionHandler> reconnectionHandlerMultibinder = Multibinder.newSetBinder(binder(), SimpleConnectionPool.ReconnectionHandler.class);
        reconnectionHandlerMultibinder.addBinding().to(OpenPaasContactConsumerReconnectionHandler.class);
    }
    ```

File: tmail-backend/tmail-third-party/openpaas/src/main/java/com/linagora/tmail/contact/OpenPaasContactsConsumer.java [hunk 2]
```diff
@@ -90,6 +90,12 @@ public void start() {
         consumeContactsDisposable = doConsumeContactMessages();
     }
 
+    public void restart() {
+        Disposable previousConsumer = consumeContactsDisposable;
+        consumeContactsDisposable = doConsumeContactMessages();
+        previousConsumer.dispose();
```
Comment (HoussemNasri) [inline tmail-backend/tmail-third-party/openpaas/src/main/java/com/linagora/tmail/contact/OpenPaasContactsConsumer.java:96]: Are we certain previousConsumer is not null?

**Review by HoussemNasri — COMMENTED** [review]
LGTM, not sure how we can write tests for this though.

**Comment by florentos17** [comment]
> LGTM, not sure how we can write tests for this though.

[tmail-backend/apps/distributed/src/test/java/com/linagora/tmail/james/app/SchedulerReconnectionHandlerIntegrationTest.java](https://github.com/linagora/tmail-backend/pull/1366/files#diff-053097d61e925b475b6cd9fb27d86e08a18e7c587c8fd2a7c9073332201a529b)

File: tmail-backend/tmail-third-party/openpaas/src/main/java/com/linagora/tmail/contact/OpenPaasContactsConsumer.java (hunk 2 above)
Comment (chibenwa) [inline tmail-backend/tmail-third-party/openpaas/src/main/java/com/linagora/tmail/contact/OpenPaasContactsConsumer.java:96]:
```suggestion
        Optional.ofNullable(previousConsumer).ifPresent(Disposable::dispose);
```

No it's null of not started yet.

**Review by chibenwa — APPROVED** [review]

**Comment by chibenwa** [comment]
Rebase needed

**Review by chibenwa — APPROVED** [review]

**Comment by chibenwa** [comment]
```
Unable to create injector, see the following errors:

1) [Guice/MissingImplementation]: No implementation for OpenPaasContactsConsumer annotated with @Named("openpaas") was bound.

Did you mean?
    * OpenPaasContactsConsumer bound at OpenPaasContactsConsumer.class(OpenPaasContactsConsumer.java:41)

Requested by:
1  : OpenPaasModule.provideReconnectionHandler(OpenPaasModule.java:104)
      \_ for 1st parameter
     at OpenPaasModule.provideReconnectionHandler(OpenPaasModule.java:104)
      \_ installed by: Modules$OverrideModule -> Modules$OverrideModule -> Modules$OverrideModule -> Modules$OverrideModule -> Modules$CombinedModule -> Modules$CombinedModule -> Modules$CombinedModule -> OpenPaasModule

Learn more:
  https://github.com/google/guice/wiki/MISSING_IMPLEMENTATION

1 error

======================
Full classname legend:
======================
Modules$CombinedModule:   "com.google.inject.util.Modules$CombinedModule"
Modules$OverrideModule:   "com.google.inject.util.Modules$OverrideModule"
Named:                    "com.google.inject.name.Named"
OpenPaasContactsConsumer: "com.linagora.tmail.contact.OpenPaasContactsConsumer"
OpenPaasModule:           "com.linagora.tmail.OpenPaasModule"
========================
End of classname legend:
========================
```

com.linagora.tmail.OpenPaasAmqpForwardAttributeIntegrationTest$WithOpenPaasModuleAndFanoutRabbitMqExchange.recipientsShouldBePublishedToAmqpWhenSendingEmailWithExchangeTypeFanoutConfiguration(File)

**Review by Arsnael — APPROVED** [review]

**Comment by Arsnael** [comment]
Need to rebase

**Comment by vttranlina** [comment]
read it

**Comment by florentos17** [comment]
i rebased and adapted the changes of the current PR to take into account the changes made [here](https://github.com/linagora/tmail-backend/pull/1367)

**Comment by chibenwa** [comment]
image


```
Assertion condition defined as a Lambda expression in com.linagora.tmail.james.app.SchedulerReconnectionHandlerIntegrationTest$DeletedMessageVaultQueueDisabled 
Multiple Failures (3 failures)
-- failure 1 --
expected: 1L
 but was: 0L
at SchedulerReconnectionHandlerIntegrationTest$DeletedMessageVaultQueueDisabled.lambda$setUp$7(SchedulerReconnectionHandlerIntegrationTest$DeletedMessageVaultQueueDisabled.java:163)
-- failure 2 --
expected: 1L
 but was: 0L
at SchedulerReconnectionHandlerIntegrationTest$DeletedMessageVaultQueueDisabled.lambda$setUp$7(SchedulerReconnectionHandlerIntegrationTest$DeletedMessageVaultQueueDisabled.java:163)
-- failure 3 --
expected: 1L
 but was: 0L
at SchedulerReconnectionHandlerIntegrationTest$DeletedMessageVaultQueueDisabled.lambda$setUp$7(SchedulerReconnectionHandlerIntegrationTest$DeletedMessageVaultQueueDisabled.java:163) within 10 seconds.
```

Looks related.

Can you check @florentos17 ?

File: tmail-backend/guice/distributed/src/main/java/com/linagora/tmail/ScheduledReconnectionHandler.java [hunk 3]
```diff
@@ -300,7 +300,10 @@ class QueueNotFoundException extends RuntimeException {
         "JamesMailQueue-workqueue-outgoing",
         "mailboxEvent-workQueue-org.apache.james.events.GroupRegistrationHandler$GroupRegistrationHandlerGroup",
         "jmapEvent-workQueue-org.apache.james.events.GroupRegistrationHandler$GroupRegistrationHandlerGroup",
-        "deleted-message-vault-work-queue");
+        "deleted-message-vault-work-queue",
+        "openpaas-contacts-queue-add",
+        "openpaas-contacts-queue-update",
+        "openpaas-contacts-queue-delete");
```
Comment (florentos17) [inline tmail-backend/guice/distributed/src/main/java/com/linagora/tmail/ScheduledReconnectionHandler.java:306]: the issue arises when the extra queues are added here.
in `SchedulerReconnectionHandlerIntegrationTest.java`, the tests expect each queue to have consumers, but these do not have any, and i don't know how to add them.
i tried to add a `OpenPaasContactsConsumerProbe.java` but it seems like it does not change much.

File: tmail-backend/apps/distributed/src/test/java/com/linagora/tmail/james/app/SchedulerReconnectionHandlerIntegrationTest.java [hunk 4]
```diff
@@ -145,6 +146,7 @@ class DeletedMessageVaultQueueDisabled {
                 .overrideWith(new LinagoraTestJMAPServerModule())
                 .overrideWith(binder -> Multibinder.newSetBinder(binder, GuiceProbe.class).addBinding().to(MailboxManagerClassProbe.class))
                 .overrideWith(binder -> Multibinder.newSetBinder(binder, GuiceProbe.class).addBinding().to(ScheduledReconnectionHandlerProbe.class))
+                //.overrideWith(binder -> Multibinder.newSetBinder(binder, GuiceProbe.class).addBinding().to(OpenPaasContactsConsumerProbe.class))
```
Comment (chibenwa) [inline tmail-backend/apps/distributed/src/test/java/com/linagora/tmail/james/app/SchedulerReconnectionHandlerIntegrationTest.java]: commented code spotted

File: tmail-backend/apps/distributed/src/test/java/com/linagora/tmail/james/app/SchedulerReconnectionHandlerIntegrationTest.java [hunk 5]
```diff
@@ -69,6 +70,7 @@ class DeletedMessageVaultQueueEnabled {
                 .build())
             .server(configuration -> DistributedServer.createServer(configuration)
                 .overrideWith(new LinagoraTestJMAPServerModule())
+                .overrideWith(new OpenPaasModule())
```
Comment (florentos17) [inline tmail-backend/apps/distributed/src/test/java/com/linagora/tmail/james/app/SchedulerReconnectionHandlerIntegrationTest.java:73]: yesterday's attempt was wrong, i dropped the commit and simply added the `OpenPaasModule`, it resolved the problem

**Review by chibenwa — APPROVED** [review]

**Comment by florentos17** [comment]
rebased on master to solve the merge conflict

**Pull request merged**